File: emisor.c
#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <arpa/inet.h>


#define MAX_MSG_LEN 256

// --- CAPA DE PRESENTACIÓN: ASCII -> binario ---
void ascii_a_binario(const char *msg, char *binario) {
    int k = 0;
    for (int i = 0; msg[i] != '\0'; i++) {
        for (int j = 7; j >= 0; j--) {
            binario[k++] = ((msg[i] >> j) & 1) ? '1' : '0';
        }
    }
    binario[k] = '\0';
}

// --- CAPA DE ENLACE: Checksum ---
void fletcher16(const char *data, int len, uint16_t *c1, uint16_t *c2) {
    *c1 = 0;
    *c2 = 0;
    // Convierte los bytes de '0'/'1' a 0/1 numérico para checksum.
    for (int i = 0; i < len; i++) {
        uint8_t val = data[i] == '1' ? 1 : 0;
        *c1 = (*c1 + val) % 255;
        *c2 = (*c2 + *c1) % 255;
    }
}

// --- CAPA DE RUIDO ---
void aplicar_ruido(char *trama, double tasa_error) {
    int len = strlen(trama);
    srand((unsigned int) time(NULL));
    for (int i = 0; i < len; i++) {
        double r = rand() / (RAND_MAX + 1.0);
        if (r < tasa_error) {
            trama[i] = (trama[i] == '1') ? '0' : '1'; // Flip bit
        }
    }
}

// --- CAPA DE APLICACIÓN ---
void solicitar_mensaje(char *mensaje, double *tasa_error) {
    printf("Ingrese el mensaje de texto: ");
    fgets(mensaje, MAX_MSG_LEN, stdin);
    mensaje[strcspn(mensaje, "\n")] = 0; // Quita el salto de línea
    printf("Ingrese la tasa de error (ejemplo 0.01 para 1%%): ");
    scanf("%lf", tasa_error);
    getchar(); // Limpia el buffer
}

void checksum_a_bin(uint16_t c1, uint16_t c2, char *checksum) {
    for (int i = 7; i >= 0; i--) checksum[7-i] = ((c1 >> i) & 1) ? '1' : '0';
    for (int i = 7; i >= 0; i--) checksum[15-i] = ((c2 >> i) & 1) ? '1' : '0';
    checksum[16] = '\0';
}

int main() {
    char mensaje[MAX_MSG_LEN];
    double tasa_error;
    solicitar_mensaje(mensaje, &tasa_error);

    // PRESENTACIÓN
    char binario[MAX_MSG_LEN * 8 + 1];
    ascii_a_binario(mensaje, binario);

    // ENLACE
    uint16_t c1, c2;
    fletcher16(binario, strlen(binario), &c1, &c2);
    char checksum[17];
    checksum_a_bin(c1, c2, checksum);

    char trama[MAX_MSG_LEN * 8 + 17];
    strcpy(trama, binario);
    strcat(trama, checksum);

    // RUIDO
    aplicar_ruido(trama, tasa_error);

    // SOCKET - TRANSMISIÓN
    int sockfd;
    struct sockaddr_in serv_addr;
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) { perror("Error creando socket"); exit(1); }
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_port = htons(5000);
    serv_addr.sin_addr.s_addr = inet_addr("127.0.0.1");

    if (connect(sockfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0) {
        perror("Error conectando al receptor");
        close(sockfd);
        exit(1);
    }
    int len = strlen(trama);
    int sent = send(sockfd, trama, len, 0);
    if (sent < 0) { perror("Error enviando"); }
    else { printf("Mensaje enviado correctamente!\n"); }
    close(sockfd);
    return 0;
}

File: emisor.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"net"
	"os"
	"strconv"
	"strings"
	"time"
)

// CAPA DE APLICACIÓN
func solicitarMensaje() (string, float64) {
	reader := bufio.NewReader(os.Stdin)
	
	fmt.Print("Ingrese el mensaje de texto: ")
	mensaje, _ := reader.ReadString('\n')
	mensaje = strings.TrimSpace(mensaje)
	
	fmt.Print("Ingrese la tasa de error (ejemplo 0.01 para 1%): ")
	errorStr, _ := reader.ReadString('\n')
	errorStr = strings.TrimSpace(errorStr)
	tasaError, _ := strconv.ParseFloat(errorStr, 64)
	
	fmt.Printf("[APPLICATION] Message: \"%s\"\n", mensaje)
	return mensaje, tasaError
}

// CAPA DE PRESENTACIÓN
func codificarMensaje(mensaje string) string {
	var binario strings.Builder
	
	for _, char := range mensaje {
		for i := 7; i >= 0; i-- {
			bit := (int(char) >> i) & 1
			binario.WriteString(strconv.Itoa(bit))
		}
	}
	
	resultado := binario.String()
	fmt.Printf("[PRESENTATION] ASCII Binary: \"%s\"\n", resultado)
	return resultado
}

// CAPA DE ENLACE - Algoritmo de Hamming
func calcularIntegridad(datos string) string {
	dataBits := len(datos)
	
	// Calcular número de bits de paridad necesarios
	parityBits := 0
	for (1 << parityBits) < (dataBits + parityBits + 1) {
		parityBits++
	}
	
	totalBits := dataBits + parityBits
	hamming := make([]int, totalBits+1) // 1-indexed
	
	// Colocar bits de datos en posiciones que no son potencias de 2
	dataIndex := 0
	for i := 1; i <= totalBits; i++ {
		if !esPotenciaDe2(i) {
			if dataIndex < len(datos) {
				hamming[i] = int(datos[dataIndex] - '0')
				dataIndex++
			}
		}
	}
	
	// Calcular bits de paridad
	for i := 0; i < parityBits; i++ {
		parityPos := 1 << i
		parity := 0
		
		for j := 1; j <= totalBits; j++ {
			if (j & parityPos) != 0 {
				parity ^= hamming[j]
			}
		}
		
		hamming[parityPos] = parity
	}
	
	// Convertir a string
	var resultado strings.Builder
	for i := 1; i <= totalBits; i++ {
		resultado.WriteString(strconv.Itoa(hamming[i]))
	}
	
	hammingStr := resultado.String()
	fmt.Printf("[LINK] With Hamming: \"%s\"\n", hammingStr)
	return hammingStr
}

func esPotenciaDe2(n int) bool {
	return n > 0 && (n&(n-1)) == 0
}

// CAPA DE RUIDO
func aplicarRuido(trama string, tasaError float64) string {
	rand.Seed(time.Now().UnixNano())
	bits := []rune(trama)
	
	erroresAplicados := 0
	for i := range bits {
		if rand.Float64() < tasaError {
			if bits[i] == '1' {
				bits[i] = '0'
			} else {
				bits[i] = '1'
			}
			erroresAplicados++
		}
	}
	
	resultado := string(bits)
	fmt.Printf("[NOISE] After noise (error rate %.3f, %d errors applied): \"%s\"\n", 
		tasaError, erroresAplicados, resultado)
	return resultado
}

// CAPA DE TRANSMISIÓN
func enviarInformacion(trama string) {
	fmt.Println("[TRANSMISSION] Sending via socket...")
	
	conn, err := net.Dial("tcp", "127.0.0.1:5000")
	if err != nil {
		fmt.Printf("Error conectando al receptor: %v\n", err)
		return
	}
	defer conn.Close()
	
	_, err = conn.Write([]byte(trama))
	if err != nil {
		fmt.Printf("Error enviando datos: %v\n", err)
		return
	}
	
	fmt.Println("Mensaje enviado correctamente!")
}

func main() {
	fmt.Println("=== EMISOR - Algoritmo de Hamming ===")
	
	// 1. CAPA DE APLICACIÓN
	mensaje, tasaError := solicitarMensaje()
	
	// 2. CAPA DE PRESENTACIÓN
	binario := codificarMensaje(mensaje)
	
	// 3. CAPA DE ENLACE
	hammingCode := calcularIntegridad(binario)
	
	// 4. CAPA DE RUIDO
	tramaConRuido := aplicarRuido(hammingCode, tasaError)
	
	// 5. CAPA DE TRANSMISIÓN
	enviarInformacion(tramaConRuido)
}
